File: app/src/main/java/wangdaye/com/geometricweather/theme/resource/providers/PixelResourcesProvider.java
package wangdaye.com.geometricweather.theme.resource.providers;

import android.animation.Animator;
import android.graphics.drawable.Drawable;
import android.net.Uri;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.Size;

import wangdaye.com.geometricweather.GeometricWeather;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.theme.resource.utils.Constants;
import wangdaye.com.geometricweather.theme.resource.utils.ResourceUtils;
import wangdaye.com.geometricweather.common.ui.images.pixel.PixelMoonDrawable;
import wangdaye.com.geometricweather.common.ui.images.pixel.PixelSunDrawable;

public class PixelResourcesProvider extends IconPackResourcesProvider {

    public PixelResourcesProvider(@NonNull ResourceProvider defaultProvider) {
        super(
                GeometricWeather.getInstance(),
                GeometricWeather.getInstance().getPackageName(),
                defaultProvider
        );
    }

    public static boolean isPixelIconProvider(@NonNull String packageName) {
        return packageName.equals(
                GeometricWeather.getInstance().getPackageName() + ".Pixel"
        );
    }

    @NonNull
    protected Uri getDrawableUri(String resName) {
        return ResourceUtils.getDrawableUri(super.getPackageName(), "drawable", resName);
    }

    @Override
    public String getPackageName() {
        return super.getPackageName() + ".Pixel";
    }

    @Override
    public String getProviderName() {
        return "Pixel";
    }

    @Override
    public Drawable getProviderIcon() {
        return getWeatherIcon(WeatherCode.PARTLY_CLOUDY, true);
    }

    // weather icon.

    @Override
    @Size(3)
    public Drawable[] getWeatherIcons(WeatherCode code, boolean dayTime) {
        return new Drawable[] {getWeatherIcon(code, dayTime), null, null};
    }

    @Override
    String getWeatherIconName(WeatherCode code, boolean daytime) {
        return super.getWeatherIconName(code, daytime) + Constants.SEPARATOR + "pixel";
    }

    @Override
    String getWeatherIconName(WeatherCode code, boolean daytime,
                                     @IntRange(from = 1, to = 3) int index) {
        if (index == 1) {
            return getWeatherIconName(code, daytime);
        } else {
            return null;
        }
    }

    // animator.

    @Override
    @Size(3)
    public Animator[] getWeatherAnimators(WeatherCode code, boolean dayTime) {
        return new Animator[] {null, null, null};
    }

    @Override
    String getWeatherAnimatorName(WeatherCode code, boolean daytime,
                                  @IntRange(from = 1, to = 3) int index) {
        return null;
    }

    // sun and moon.

    @NonNull
    public Drawable getSunDrawable() {
        return new PixelSunDrawable();
    }

    @NonNull
    public Drawable getMoonDrawable() {
        return new PixelMoonDrawable();
    }

    @Override
    @Nullable
    String getSunDrawableClassName() {
        return PixelSunDrawable.class.toString();
    }

    @Override
    @Nullable
    String getMoonDrawableClassName() {
        return PixelMoonDrawable.class.toString();
    }
}

File: app/src/main/java/org/breezyweather/theme/resource/providers/PixelResourcesProvider.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.theme.resource.providers

import android.animation.Animator
import android.graphics.drawable.Drawable
import android.net.Uri
import androidx.annotation.IntRange
import androidx.annotation.Size
import org.breezyweather.BreezyWeather
import breezyweather.domain.weather.model.WeatherCode
import org.breezyweather.common.ui.images.pixel.PixelMoonDrawable
import org.breezyweather.common.ui.images.pixel.PixelSunDrawable
import org.breezyweather.theme.resource.utils.Constants
import org.breezyweather.theme.resource.utils.ResourceUtils

class PixelResourcesProvider(defaultProvider: ResourceProvider) : IconPackResourcesProvider(
    BreezyWeather.instance,
    BreezyWeather.instance.packageName,
    defaultProvider
) {
    override fun getDrawableUri(resName: String): Uri {
        return ResourceUtils.getDrawableUri(super.packageName, "drawable", resName)
    }

    override val packageName: String
        get() = super.packageName + ".Pixel"
    override var providerName: String?
        get() = "Pixel"
        set(providerName) {
            super.providerName = providerName
        }
    override val providerIcon: Drawable
        get() = getWeatherIcon(WeatherCode.PARTLY_CLOUDY, true)

    // weather icon.
    @Size(3)
    override fun getWeatherIcons(code: WeatherCode?, dayTime: Boolean): Array<Drawable?> {
        return arrayOf(getWeatherIcon(code, dayTime), null, null)
    }

    override fun getWeatherIconName(code: WeatherCode?, daytime: Boolean): String {
        return super.getWeatherIconName(code, daytime) + Constants.SEPARATOR + "pixel"
    }

    override fun getWeatherIconName(
        code: WeatherCode?, daytime: Boolean,
        @IntRange(from = 1, to = 3) index: Int
    ): String? {
        return if (index == 1) {
            getWeatherIconName(code, daytime)
        } else null
    }

    // animator.
    @Size(3)
    override fun getWeatherAnimators(code: WeatherCode?, dayTime: Boolean): Array<Animator?> {
        return arrayOf(null, null, null)
    }

    override fun getWeatherAnimatorName(
        code: WeatherCode?, daytime: Boolean,
        @IntRange(from = 1, to = 3) index: Int
    ): String? {
        return null
    }

    override val sunDrawable: Drawable
        // sun and moon.
        get() = PixelSunDrawable()
    override val moonDrawable: Drawable
        get() = PixelMoonDrawable()
    override val sunDrawableClassName: String
        get() = PixelSunDrawable::class.java.toString()
    override val moonDrawableClassName: String
        get() = PixelMoonDrawable::class.java.toString()

    companion object {
        fun isPixelIconProvider(packageName: String): Boolean {
            return packageName == BreezyWeather.instance.packageName + ".Pixel"
        }
    }
}
